### PR TITLE
[4.0] [a11y] aria-describedby when field has a description

### PR DESCRIPTION
**Pull request by brianteeman**
The aria-described-by was in the wrong place and there was an unneeded title attribute

This was reported by @zwiastunsw #23896

I am working through the fields

fields included so far
- list
- number
- text
- textarea
- calendar
- url
- tel
- email
- range
- accesslevel
- combo

NOTE - this change will need to be done for all field layouts even if core doesnt have any cases where there is a description

### CODE REVIEW

**Comment by dgrammatiko** [comment]
@brianteeman one thing to consider here: this description in J3 used to be the tooltip of each field. Some texts are way too long and probably inappropriate for an `aria-describedby`

Maybe for the shake of B/C we should leave this as is (and use it in the new tooltips, activated on mouse over a field, or on field focus) and introduce another attribute, eg `describe` ?

FWIW I'm not against what you're doing here, just thinking what will be the less painful approach here...

**Comment by brianteeman** [comment]
@dgrammatiko I would much prefer it if peo0ple could SEE those crazy long tips and submit pr to change them to something other than a manual

**Comment by dgrammatiko** [comment]
@brianteeman and we both know that some will resist those changes...

**Comment by brianteeman** [comment]
@dgrammatiko https://www.youtube.com/watch?v=rtEaR1JU-ps

**Comment by zwiastunsw** [comment]
After the changes the descriptions are read by the screen reader. Perfectly.
@dgrammatiko : The number of characters in the aria-describedby attribute is not limited. But I agree that some descriptions need to be shortened.

**Comment by zwiastunsw** [comment]
I have tested this item :white_check_mark: successfully on e38aaa455222b0e60baf0b815feca10397ba2876<br /><br />I test pages: Article new/edit, Menus new/edit, Modules new/edit. Different field types <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23919">issues.joomla.org/tracker/joomla-cms/23919</a>.</sub>

**Comment by zwiastunsw** [comment]
@brianteeman Please, correct PR title: `aria-describedby`

**Comment by brianteeman** [comment]
@zwiastunsw Thanks for testing the work so far. I will continue on this tomorrow to make sure that all fields have this functionality even if the core doesnt use it

File: layouts/joomla/form/field/list.php [hunk 1]
```diff
@@ -55,6 +56,8 @@
 $attr .= $required ? ' required' : '';
 $attr .= $autofocus ? ' autofocus' : '';
 $attr .= $onchange ? ' onchange="' . $onchange . '"' : '';
+$attr .= (!empty($description)) ? ' aria-describedby="' . $aria . '"' : '';
```
Comment (Quy) [inline layouts/joomla/form/field/list.php]:
```suggestion
$attr .= !empty($description) ? ' aria-describedby="' . $aria . '"' : '';
```

**Comment by brianteeman** [comment]
@dgrammatiko 
> Maybe for the shake of B/C we should leave this as is (and use it in the new tooltips, activated on mouse over a field, or on field focus) and introduce another attribute, eg `describe` ?

Are these "new tooltips" available now? We need to resolve the search box descriptions that previously looked like
image

**Comment by dgrammatiko** [comment]
@brianteeman the code needs some cleanup and fixing the positioning: https://codepen.io/dgrammatiko/pen/eGELjo

**Comment by brianteeman** [comment]
@dgrammatiko looks good. We need this for the search tips. If you create a pr that would be awesome

**Comment by dgrammatiko** [comment]
@brianteeman cans you please cancel the new variable?
Eg: this
```
$aria = $name . '-desc';
	$attr .= (!empty($description)) ? ' aria-describedby="' . $aria . '"' : '';
```
becomes
```
	$attr .= (!empty($description)) ? ' aria-describedby="' . $name . '-desc' . '"' : '';
```

Rule of thumb: declare a new variable only if you're gonna use it more than once, else you just adding more memory allocation for the app

**Comment by zwiastunsw** [comment]
We also need it for buttons - font icons (without label).

**Comment by brianteeman** [comment]
updated pr as requested

**Comment by zwiastunsw** [comment]
@brianteeman : 
I found two cases where the description is not announced by the screen reader:
* switcher case (this will be a separate PR)
* control type : `joomla-field-fancy-select`
tagi

Page type: Module site new/edit - Article Category, Popular Tags, Similar Tags
In all other cases - it is OK.

File: layouts/joomla/form/field/accesslevel.php [hunk 2]
```diff
@@ -55,5 +55,6 @@
 $attr .= $required ? ' required' : '';
 $attr .= $autofocus ? ' autofocus' : '';
 $attr .= $onchange ? ' onchange="' . $onchange . '"' : '';
+$attr .= (!empty($description)) ? ' aria-describedby="' . $name . '-desc"' : '';
```
Comment (Quy) [inline layouts/joomla/form/field/accesslevel.php]:
```suggestion
$attr .= !empty($description) ? ' aria-describedby="' . $name . '-desc"' : '';
```

File: layouts/joomla/form/field/combo.php [hunk 3]
```diff
@@ -52,6 +52,7 @@
 $attr .= !empty($readonly) ? ' readonly' : '';
 $attr .= !empty($disabled) ? ' disabled' : '';
 $attr .= !empty($required) ? ' required' : '';
+$attr .= (!empty($description)) ? ' aria-describedby="' . $name . '-desc"' : '';
```
Comment (Quy) [inline layouts/joomla/form/field/combo.php]:
```suggestion
$attr .= !empty($description) ? ' aria-describedby="' . $name . '-desc"' : '';
```

**Comment by brianteeman** [comment]
> * control type : `joomla-field-fancy-select`

This is another javascript one so it will have to be dealt with in a different pr by @dgrammatiko

File: layouts/joomla/form/field/list.php [hunk 4]
```diff
@@ -55,6 +55,8 @@
 $attr .= $required ? ' required' : '';
 $attr .= $autofocus ? ' autofocus' : '';
 $attr .= $onchange ? ' onchange="' . $onchange . '"' : '';
+$attr .= !empty($description) ? ' aria-describedby="' . $name . '-desc"' : '';
+
```
Comment (Quy) [inline layouts/joomla/form/field/list.php:59]:
```suggestion
```

File: layouts/joomla/form/field/email.php [hunk 5]
```diff
@@ -53,6 +53,7 @@
 $attributes = array(
 	$spellcheck ? '' : 'spellcheck="false"',
 	!empty($size) ? 'size="' . $size . '"' : '',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/email.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

File: layouts/joomla/form/field/number.php [hunk 6]
```diff
@@ -49,6 +49,7 @@
 
 $attributes = array(
 	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/number.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

File: layouts/joomla/form/field/range.php [hunk 7]
```diff
@@ -45,8 +45,10 @@
  */
 
 // Initialize some field attributes.
+
 $attributes = array(
 	$class ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/range.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

File: layouts/joomla/form/field/tel.php [hunk 8]
```diff
@@ -50,6 +50,7 @@
 
 $attributes = array(
 	!empty($size) ? 'size="' . $size . '"' : '',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/tel.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

File: layouts/joomla/form/field/text.php [hunk 9]
```diff
@@ -59,7 +59,7 @@
 $attributes = array(
 	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
 	!empty($size) ? 'size="' . $size . '"' : '',
-	!empty($description) ? 'title="' . $description . '"' : '',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/text.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

File: layouts/joomla/form/field/textarea.php [hunk 10]
```diff
@@ -52,6 +52,7 @@
 	$columns ?: '',
 	$rows ?: '',
 	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!empty($description) ? ' aria-describedby="' . $name . '-desc"' : '',
```
Comment (Quy) [inline layouts/joomla/form/field/textarea.php]:
```suggestion
	!empty($description) ? 'aria-describedby="' . $name . '-desc"' : '',
```

**Comment by Quy** [comment]
When in an array, no need to add a leading space since it will be imploded later with a space.

This should also be fixed in `layouts/joomla/form/field/url.php` with all the attributes.

**Comment by brianteeman** [comment]
@wilsonge can you merge this one as well please. There won't be any more fields added to this PR as the remaining ones require js

**Comment by wilsonge** [comment]
I don't have permission to your repo to merge in 4.0. so yes - but i need you to sync up with 4.0 please

**Comment by brianteeman** [comment]
@wilsonge should be ok now

**Comment by wilsonge** [comment]
waiting for rips to be sorted but intention is to merge this next

**Comment by brianteeman** [comment]
OK - thanks

**Comment by wilsonge** [comment]
Thanks!

**Comment by brianteeman** [comment]
Thanks

**Comment by Quy** [comment]
How to make radio buttons a11y or is it already as is? There is id `jform[sticky]-desc` but no reference to it.

```
<div class="control-group">
  <div class="control-label sr-only">
    <label id="jform_sticky-lbl" > Pinned</label>
  </div>
  <div class="controls">
    <fieldset>
      <legend class="switcher__legend"> Pinned </legend>
      <div class="switcher">
        <input type="radio" id="jform_sticky0" name="jform[sticky]" value="0" checked="checked" class="active">
        <label for="jform_sticky0">No</label>
        <input type="radio" id="jform_sticky1" name="jform[sticky]" value="1" >
        <label for="jform_sticky1">Yes</label>
        <span class="toggle-outside"><span class="toggle-inside"></span></span> </div>
    </fieldset>
  </div>
  <div id="jform[sticky]-desc"> <small class="form-text text-muted"> Whether or not the Banner is 'pinned'. If one or more Banners in a Category are pinned, they will take priority over Banners that are not pinned. For example, if two Banners in a Category are pinned and a third Banner is not pinned, the third Banner will not display if the module setting is 'Pinned, Randomise'. Only the two pinned Banners will display. </small> </div>
```

**Pull request merged**